File: cpp/exercises/practice/unit_test_d456ce73_603d_4e0c_a42e_69addaedfb89/distributed_tx/distributed_tx_test.cpp
#include "catch.hpp"
#include "distributed_tx.h"
#include <chrono>
#include <future>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>
#include <vector>

class MockService {
public:
    MockService(const std::string& name, bool willSucceed = true, bool willThrow = false, 
                std::chrono::milliseconds prepareDelay = std::chrono::milliseconds(0),
                std::chrono::milliseconds commitDelay = std::chrono::milliseconds(0),
                std::chrono::milliseconds rollbackDelay = std::chrono::milliseconds(0),
                int failAfterCommitAttempts = -1)
        : name_(name), willSucceed_(willSucceed), willThrow_(willThrow),
          prepareDelay_(prepareDelay), commitDelay_(commitDelay), rollbackDelay_(rollbackDelay),
          failAfterCommitAttempts_(failAfterCommitAttempts), 
          prepareCount_(0), commitCount_(0), rollbackCount_(0) {}

    bool prepare() {
        ++prepareCount_;
        if (prepareDelay_.count() > 0) {
            std::this_thread::sleep_for(prepareDelay_);
        }
        if (willThrow_) {
            throw std::runtime_error(name_ + " prepare failed with exception");
        }
        return willSucceed_;
    }

    void commit() {
        ++commitCount_;
        if (commitDelay_.count() > 0) {
            std::this_thread::sleep_for(commitDelay_);
        }
        if (failAfterCommitAttempts_ >= 0 && commitCount_ <= failAfterCommitAttempts_) {
            throw std::runtime_error(name_ + " commit failed with exception");
        }
    }

    void rollback() {
        ++rollbackCount_;
        if (rollbackDelay_.count() > 0) {
            std::this_thread::sleep_for(rollbackDelay_);
        }
    }

    int getPrepareCount() const { return prepareCount_; }
    int getCommitCount() const { return commitCount_; }
    int getRollbackCount() const { return rollbackCount_; }
    std::string getName() const { return name_; }

private:
    std::string name_;
    bool willSucceed_;
    bool willThrow_;
    std::chrono::milliseconds prepareDelay_;
    std::chrono::milliseconds commitDelay_;
    std::chrono::milliseconds rollbackDelay_;
    int failAfterCommitAttempts_;
    int prepareCount_;
    int commitCount_;
    int rollbackCount_;
};

TEST_CASE("Successful transaction with all services succeeding", "[transaction]") {
    auto service1 = std::make_shared<MockService>("service1");
    auto service2 = std::make_shared<MockService>("service2");
    auto service3 = std::make_shared<MockService>("service3");

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    REQUIRE(result);
    REQUIRE(service1->getPrepareCount() == 1);
    REQUIRE(service2->getPrepareCount() == 1);
    REQUIRE(service3->getPrepareCount() == 1);
    REQUIRE(service1->getCommitCount() == 1);
    REQUIRE(service2->getCommitCount() == 1);
    REQUIRE(service3->getCommitCount() == 1);
    REQUIRE(service1->getRollbackCount() == 0);
    REQUIRE(service2->getRollbackCount() == 0);
    REQUIRE(service3->getRollbackCount() == 0);
}

TEST_CASE("Transaction rolled back when a service returns false for prepare", "[transaction]") {
    auto service1 = std::make_shared<MockService>("service1");
    auto service2 = std::make_shared<MockService>("service2", false); // Will return false for prepare
    auto service3 = std::make_shared<MockService>("service3");

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    REQUIRE_FALSE(result);
    REQUIRE(service1->getPrepareCount() >= 1); // May not call prepare on all services if using parallelization
    REQUIRE(service2->getPrepareCount() == 1);
    REQUIRE(service1->getCommitCount() == 0);
    REQUIRE(service2->getCommitCount() == 0);
    REQUIRE(service3->getCommitCount() == 0);
    REQUIRE(service1->getRollbackCount() == 1);
    REQUIRE(service2->getRollbackCount() == 1);
    REQUIRE(service3->getRollbackCount() == 1);
}

TEST_CASE("Transaction rolled back when a service throws an exception during prepare", "[transaction]") {
    auto service1 = std::make_shared<MockService>("service1");
    auto service2 = std::make_shared<MockService>("service2", true, true); // Will throw during prepare
    auto service3 = std::make_shared<MockService>("service3");

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    REQUIRE_FALSE(result);
    REQUIRE(service2->getPrepareCount() == 1);
    REQUIRE(service1->getCommitCount() == 0);
    REQUIRE(service2->getCommitCount() == 0);
    REQUIRE(service3->getCommitCount() == 0);
    REQUIRE(service1->getRollbackCount() == 1);
    REQUIRE(service2->getRollbackCount() == 1);
    REQUIRE(service3->getRollbackCount() == 1);
}

TEST_CASE("Transaction timeout during prepare phase", "[transaction]") {
    auto service1 = std::make_shared<MockService>("service1");
    // This service will take too long to prepare
    auto service2 = std::make_shared<MockService>("service2", true, false, std::chrono::milliseconds(6000));
    auto service3 = std::make_shared<MockService>("service3");

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.set_prepare_timeout(std::chrono::seconds(1)); // Set a short timeout for testing
    coordinator.begin_transaction(services);
    
    bool result = coordinator.execute_transaction();

    REQUIRE_FALSE(result);
    // Service2 might not complete prepare due to timeout
    REQUIRE(service1->getCommitCount() == 0);
    REQUIRE(service2->getCommitCount() == 0);
    REQUIRE(service3->getCommitCount() == 0);
    REQUIRE(service1->getRollbackCount() == 1);
    REQUIRE(service2->getRollbackCount() == 1);
    REQUIRE(service3->getRollbackCount() == 1);
}

TEST_CASE("Retry mechanism for commit phase", "[transaction]") {
    auto service1 = std::make_shared<MockService>("service1");
    // This service will fail the first commit attempt but succeed on second attempt
    auto service2 = std::make_shared<MockService>("service2", true, false, 
                                                std::chrono::milliseconds(0), 
                                                std::chrono::milliseconds(0), 
                                                std::chrono::milliseconds(0), 
                                                1); // Fail after 1 commit attempt
    auto service3 = std::make_shared<MockService>("service3");

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.set_commit_max_retries(3); // Allow up to 3 retries
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    REQUIRE(result);
    REQUIRE(service1->getPrepareCount() == 1);
    REQUIRE(service2->getPrepareCount() == 1);
    REQUIRE(service3->getPrepareCount() == 1);
    REQUIRE(service1->getCommitCount() == 1);
    REQUIRE(service2->getCommitCount() > 1); // Should have been called multiple times due to retries
    REQUIRE(service3->getCommitCount() == 1);
    REQUIRE(service1->getRollbackCount() == 0);
    REQUIRE(service2->getRollbackCount() == 0);
    REQUIRE(service3->getRollbackCount() == 0);
}

TEST_CASE("Transaction with large number of services", "[transaction][performance]") {
    const int N = 100; // Large enough to test performance but not too large for unit tests
    std::vector<std::shared_ptr<MockService>> services;
    
    for (int i = 0; i < N; ++i) {
        services.push_back(std::make_shared<MockService>("service" + std::to_string(i)));
    }

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    
    auto start = std::chrono::high_resolution_clock::now();
    bool result = coordinator.execute_transaction();
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double, std::milli> elapsed = end - start;
    
    REQUIRE(result);
    
    // Check that all services were called correctly
    for (auto& service : services) {
        REQUIRE(service->getPrepareCount() == 1);
        REQUIRE(service->getCommitCount() == 1);
        REQUIRE(service->getRollbackCount() == 0);
    }
    
    // Optional performance check - this depends on your implementation and hardware
    // INFO("Transaction with " << N << " services completed in " << elapsed.count() << "ms");
}

TEST_CASE("Concurrent service preparation", "[transaction][concurrency]") {
    // Services with significant delays to test concurrent preparation
    auto service1 = std::make_shared<MockService>("service1", true, false, std::chrono::milliseconds(300));
    auto service2 = std::make_shared<MockService>("service2", true, false, std::chrono::milliseconds(300));
    auto service3 = std::make_shared<MockService>("service3", true, false, std::chrono::milliseconds(300));

    std::vector<std::shared_ptr<MockService>> services = {service1, service2, service3};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    
    auto start = std::chrono::high_resolution_clock::now();
    bool result = coordinator.execute_transaction();
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double, std::milli> elapsed = end - start;
    
    REQUIRE(result);
    
    // If preparation is done concurrently, the total time should be closer to the max delay
    // of a single service rather than the sum of all delays
    // Note: This test might fail if concurrency is not implemented or if the system is heavily loaded
    INFO("Transaction completed in " << elapsed.count() << "ms");
    // We don't make a hard requirement here, as it depends on the implementation
    // but we can check: 
    // REQUIRE(elapsed.count() < 600); // Less than the sum of delays (900ms)
}

TEST_CASE("Empty service list", "[transaction][edge_case]") {
    std::vector<std::shared_ptr<MockService>> services;

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    // Empty transaction should succeed (no services to commit or roll back)
    REQUIRE(result);
}

TEST_CASE("Idempotent rollback", "[transaction]") {
    auto service = std::make_shared<MockService>("service", false); // Will fail on prepare

    std::vector<std::shared_ptr<MockService>> services = {service};

    distributed_tx::TransactionCoordinator coordinator;
    coordinator.begin_transaction(services);
    bool result = coordinator.execute_transaction();

    REQUIRE_FALSE(result);
    REQUIRE(service->getPrepareCount() == 1);
    REQUIRE(service->getCommitCount() == 0);
    REQUIRE(service->getRollbackCount() == 1); // Rollback should be called exactly once
}

File: java/exercises/practice/unit_test_ddc7dee1_b058_4db6_9fbb_9b1ffda44d97/distributed_tx/src/test/java/DistributedTxTest.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public class DistributedTxTest {

    class MockService implements Service {
        private final String name;
        private final boolean prepareSuccess;
        private final boolean commitSuccess;
        private final boolean throwPrepareException;
        private final boolean throwCommitException;
        private int prepareCount = 0;
        private int commitCount = 0;

        MockService(String name, boolean prepareSuccess, boolean commitSuccess, 
                   boolean throwPrepareException, boolean throwCommitException) {
            this.name = name;
            this.prepareSuccess = prepareSuccess;
            this.commitSuccess = commitSuccess;
            this.throwPrepareException = throwPrepareException;
            this.throwCommitException = throwCommitException;
        }

        @Override
        public boolean prepare() throws Exception {
            prepareCount++;
            if (throwPrepareException) {
                throw new Exception("Prepare failed for " + name);
            }
            return prepareSuccess;
        }

        @Override
        public boolean commit() throws Exception {
            commitCount++;
            if (throwCommitException) {
                throw new Exception("Commit failed for " + name);
            }
            return commitSuccess;
        }
    }

    @Test
    public void testSuccessfulTransaction() throws Exception {
        List<Service> services = new ArrayList<>();
        services.add(new MockService("Service1", true, true, false, false));
        services.add(new MockService("Service2", true, true, false, false));
        services.add(new MockService("Service3", true, true, false, false));

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertTrue(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(1, mock.commitCount);
        }
    }

    @Test
    public void testPrepareFailure() throws Exception {
        List<Service> services = new ArrayList<>();
        services.add(new MockService("Service1", true, true, false, false));
        services.add(new MockService("Service2", false, true, false, false)); // Will fail prepare
        services.add(new MockService("Service3", true, true, false, false));

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertFalse(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(0, mock.commitCount); // No commit should be called
        }
    }

    @Test
    public void testPrepareException() throws Exception {
        List<Service> services = new ArrayList<>();
        services.add(new MockService("Service1", true, true, false, false));
        services.add(new MockService("Service2", true, true, true, false)); // Will throw on prepare
        services.add(new MockService("Service3", true, true, false, false));

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertFalse(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(0, mock.commitCount); // No commit should be called
        }
    }

    @Test
    public void testCommitFailure() throws Exception {
        List<Service> services = new ArrayList<>();
        services.add(new MockService("Service1", true, true, false, false));
        services.add(new MockService("Service2", true, false, false, false)); // Will fail commit
        services.add(new MockService("Service3", true, true, false, false));

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertFalse(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(1, mock.commitCount);
        }
    }

    @Test
    public void testCommitException() throws Exception {
        List<Service> services = new ArrayList<>();
        services.add(new MockService("Service1", true, true, false, false));
        services.add(new MockService("Service2", true, true, false, true)); // Will throw on commit
        services.add(new MockService("Service3", true, true, false, false));

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertFalse(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(1, mock.commitCount);
        }
    }

    @Test
    public void testEmptyServicesList() throws Exception {
        List<Service> services = new ArrayList<>();
        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertTrue(result); // Empty transaction should succeed
    }

    @Test
    public void testLargeNumberOfServices() throws Exception {
        List<Service> services = new ArrayList<>();
        for (int i = 0; i < 1000; i++) {
            services.add(new MockService("Service" + i, true, true, false, false));
        }

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertTrue(result);
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(1, mock.commitCount);
        }
    }

    @Test
    public void testConcurrentExecution() throws Exception {
        List<Service> services = new ArrayList<>();
        AtomicBoolean prepareStarted = new AtomicBoolean(false);
        AtomicBoolean prepareCompleted = new AtomicBoolean(false);

        // Add services that will coordinate to test concurrency
        for (int i = 0; i < 10; i++) {
            final int serviceNum = i;
            services.add(new MockService("Service" + i, true, true, false, false) {
                @Override
                public boolean prepare() throws Exception {
                    if (serviceNum == 0) prepareStarted.set(true);
                    while (serviceNum == 5 && !prepareStarted.get()) {
                        Thread.sleep(10);
                    }
                    if (serviceNum == 9) prepareCompleted.set(true);
                    return super.prepare();
                }
            });
        }

        DistributedTx dtc = new DistributedTx();
        boolean result = dtc.transact(services);
        
        assertTrue(result);
        assertTrue(prepareStarted.get());
        assertTrue(prepareCompleted.get());
        for (Service service : services) {
            MockService mock = (MockService) service;
            assertEquals(1, mock.prepareCount);
            assertEquals(1, mock.commitCount);
        }
    }
}